File: app/src/main/java/cn/nicolite/huthelper/presenter/FeedBackPresenter.java
package cn.nicolite.huthelper.presenter;

import android.os.Build;
import android.text.TextUtils;

import cn.nicolite.huthelper.BuildConfig;
import cn.nicolite.huthelper.base.BasePresenter;
import cn.nicolite.huthelper.model.bean.User;
import cn.nicolite.huthelper.network.APIUtils;
import cn.nicolite.huthelper.network.exception.ExceptionEngine;
import cn.nicolite.huthelper.view.activity.FeedBackActivity;
import cn.nicolite.huthelper.view.iview.IFeedBackView;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import okhttp3.ResponseBody;

/**
 * Created by nicolite on 17-10-24.
 */

public class FeedBackPresenter extends BasePresenter<IFeedBackView, FeedBackActivity> {

    public FeedBackPresenter(IFeedBackView view, FeedBackActivity activity) {
        super(view, activity);
    }

    public void feeBack(String content, String contact) {

        User user = configure.getUser();

        if (TextUtils.isEmpty(content)) {
            if (getView() != null) {
                getView().showMessage("反馈意见不能为空！");
            }
        } else if (contact.length() > 200) {
            if (getView() != null) {
                getView().showMessage("字数超过限制！");
            }
        } else if (TextUtils.isEmpty(contact)) {
            if (getView() != null) {
                getView().showMessage("联系方式不能为空！");
            }
        } else {
            String version = "版本：Android " + BuildConfig.VERSION_NAME + "（" + BuildConfig.VERSION_CODE + "）";
            String model = "机型：" + Build.MANUFACTURER + Build.MODEL + "（Android " + Build.VERSION.RELEASE + "）";
            String from = "来源：" + user.getTrueName() + " " + user.getClass_name() + " " + user.getStudentKH();

            if (!TextUtils.isEmpty(version) && !TextUtils.isEmpty(model)) {
                content = from + "<br/>" + version + "<br/>" + model + "<br/>" + "内容：" + content;
            }

            APIUtils.INSTANCE
                    .getFeedBackAPI()
                    .feedBack(contact, content)
                    .compose(getActivity().<ResponseBody>bindToLifecycle())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Observer<ResponseBody>() {
                        @Override
                        public void onSubscribe(@NonNull Disposable d) {
                            if (getView() != null) {
                                getView().showLoading();
                            }
                        }

                        @Override
                        public void onNext(@NonNull ResponseBody responseBody) {
                            if (getView() != null) {
                                getView().closeLoading();
                                getView().onSuccess();
                            }

                        }

                        @Override
                        public void onError(@NonNull Throwable e) {
                            if (getView() != null) {
                                getView().closeLoading();
                                getView().showMessage(ExceptionEngine.handleException(e).getMsg());
                            }

                        }

                        @Override
                        public void onComplete() {

                        }
                    });
        }
    }
}

File: app/src/main/java/cn/nicolite/huthelper/view/presenter/FeedBackPresenter.kt
package cn.nicolite.huthelper.view.presenter

import android.content.pm.PackageManager
import android.os.Build
import android.text.TextUtils
import cn.nicolite.huthelper.base.BasePresenter
import cn.nicolite.huthelper.network.APIUtils
import cn.nicolite.huthelper.network.exception.ExceptionEngine
import cn.nicolite.huthelper.view.activity.FeedBackActivity
import cn.nicolite.huthelper.view.iview.IFeedBackView
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.annotations.NonNull
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody

/**
 * Created by nicolite on 17-10-24.
 */

class FeedBackPresenter(view: IFeedBackView, activity: FeedBackActivity) : BasePresenter<IFeedBackView, FeedBackActivity>(view, activity) {

    fun feeBack(content: String, contact: String) {
        val user = configure.user
        when {
            TextUtils.isEmpty(content) -> getView()?.showMessage("反馈意见不能为空！")
            contact.length > 200 -> getView()?.showMessage("字数超过限制！")
            TextUtils.isEmpty(contact) -> getView()?.showMessage("联系方式不能为空！")
            else -> {
                var version = ""
                var model = ""
                try {
                    val pm = activity.packageManager
                    val pi = pm.getPackageInfo(activity.packageName, PackageManager.GET_ACTIVITIES)
                    version = "版本：Android " + pi.versionName + "（" + pi.versionCode + "）"
                    model = "机型：" + Build.MANUFACTURER + Build.MODEL + "（Android " + Build.VERSION.RELEASE + "）"
                } catch (e: PackageManager.NameNotFoundException) {
                    e.printStackTrace()
                }

                val from = "来源：" + user.trueName + " " + user.class_name + " " + user.studentKH

                val finalContent = "$from<br/>$version<br/>$model<br/>内容：$content"

                APIUtils
                        .getFeedBackAPI()
                        .feedBack()
                        .compose(activity.bindToLifecycle())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(object : Observer<ResponseBody> {
                            override fun onSubscribe(@NonNull d: Disposable) {
                                getView()?.showLoading()
                            }

                            override fun onNext(@NonNull responseBody: ResponseBody) {
                                getView()?.closeLoading()
                                getView()?.onSuccess()
                            }

                            override fun onError(@NonNull e: Throwable) {
                                getView()?.closeLoading()
                                getView()?.showMessage(ExceptionEngine.handleException(e).msg)
                            }

                            override fun onComplete() {

                            }
                        })
            }
        }
    }
}
